### PR TITLE
feat(sdk): add bim.structural query namespace over the structural analysis read model

### PR DESCRIPTION
**Pull request by BIMvoice**
Layer 3 of the six in #4206 — the query surface. **#4206 stays open after this PR**; see "What remains" below.

## History this builds on

#4510 (merged) shipped layer 2 — `extractStructuralOnDemand(store)`, a read model of analysis models, members, connections, activities, load groups and result groups, cross-linked by GlobalId — deliberately wired into nothing so it could land alone. Its own PR body proposed the remaining slices and their order; the issue comments on #4206 repeat that plan and note it as "Depends on #4510 only". This PR is slice 1 of that list: the query surface.

The maintainer has not replied on #4206 since that comment (last activity 2026-09-11); there is no newer binding decision to follow or route around.

## Why this slice

Per the task brief for this stack: prefer the layer that makes the read model observable to a consumer, since layer 2 currently sits wired to nothing. A `bim.structural` namespace is the smallest change that gives the CLI, MCP, SDK scripts and the viewer's local backend a way to actually read the extraction, mirroring how `bim.schedule` already exposes `extractScheduleOnDemand`.

## What this adds

- `packages/sdk/src/namespaces/structural.ts` — `StructuralNamespace`: `data(modelId?)` (the full extraction plus `loadsTruncated`), and `analysisModels()`, `members()`, `connections()`, `activities()`, `loadGroups()`, `resultGroups()` as convenience accessors over the same collections.
- `packages/sandbox/src/bridge-structural.ts` — the sandbox script bridge (`bim.structural.*` inside QuickJS), added to `NAMESPACE_SCHEMAS` in `bridge-schema.ts`.
- `createStructuralAdapter()` wired into both headless backends (`packages/cli/src/headless-backend.ts`, `packages/mcp/src/headless-backend.ts`) and the viewer's `LocalBackend` (`apps/viewer/src/sdk/adapters/structural-adapter.ts`, `apps/viewer/src/sdk/local-backend.ts`) — needed so `BimBackend` stays satisfiable everywhere `bim.schedule` already is; no UI was added.
- `pnpm generate:bim-globals` regenerated and committed (`apps/viewer/src/lib/scripts/templates/bim-globals.d.ts`).
- `pnpm api-surface:update` committed — 14 new SDK export lines, all additive.

**`truncated` / `loadsTruncated` is forwarded, never dropped.** #4510 added `configuration.truncated` (a load reader bound was hit) and `StructuralExtraction.loadsTruncated` (derived from it) specifically so a consumer can't report partial load data as complete. This namespace does not recompute or default that flag anywhere in the chain — extractor → `createStructuralAdapter` (CLI/MCP/viewer) → `StructuralNamespace.data()` → `bridge-structural.ts`'s `translateData`/`translateConfiguration` all forward it unchanged. A mutation that hardcoded either flag to `false` at the bridge layer was caught by the test suite (see below).

**No parallel arrays.** Everywhere this PR pairs two things (a load configuration's entries, a connection's applied condition against its DOFs) it reuses #4510's existing paired-record shapes rather than introducing a new pair of indexed arrays.

**Attribute access.** This layer reads no raw IFC attributes itself — it only reshapes what `extractStructuralOnDemand` already resolved via the generated registry. Nothing here adds a positional index.

## Evidence

**RED**: before `structural.ts` existed, the new tests failed with `Cannot find module './structural.js'`.

**GREEN**: `packages/sdk` 225/225, `packages/sandbox` 217/217, `packages/cli` 961/961, `packages/mcp` 406/406 — all re-run and confirmed locally, not just claimed.

**Non-vacuous, against the real fixture.** `packages/cli/src/headless-backend-structural.test.ts` runs `bim.structural` through the real `HeadlessBackend`/`BimContext` wiring against `tests/models/ifcopenshell/structural_analysis_curve.ifc` and asserts specific values from that file: 1 analysis model, 3 curve members, 4 point connections, 10 activities (1 action + 9 reactions), 1 load case, 1 result group; the continuous beam's member↔connection degree sequence `[1,1,2,2]`; the applied load's `LinearForceZ: -100` at locations `[[96],[192]]`; both fixed supports as six boolean-true DOFs; and `loadsTruncated === false` for this (small, complete) fixture, distinguishing it from a truncated case.

**Mutation testing** — 4 mutations, each carrying a `console.error` hit-probe confirming the mutated line actually ran, each reverted after scoring:

| mutation | probe hit | tests reddened |
| --- | --- | --- |
| `bridge-structural.ts` `translateConfiguration`: hardcode `Truncated: false` | yes | 1 |
| `bridge-structural.ts` `translateData`: hardcode `LoadsTruncated: false` | yes | 1 |
| `StructuralNamespace.members()` rerouted to return `connections` | yes (×2) | 2 |
| MCP `createStructuralAdapter`'s `assertKnownModelId` guard dropped | yes | 1 (extended `headless-backend-model-id.test.ts`'s existing "one backend, one answer" invariant to cover `bim.structural`) |

## Gates run locally

`pnpm typecheck` (2033 test files, 51 packages, clean), `pnpm lint` (0 errors — 5 pre-existing warnings unrelated to this diff), `node scripts/check-module-size.mjs` (TS side clean; three existing rows raised to their new measured counts — `cli/headless-backend.ts` 822, `sandbox/bridge-schema.ts` 502, `sdk/types.ts` 1048 — and one new row added for `mcp/headless-backend.ts` at 424; the script's remaining non-zero exit is a pre-existing Rust allowlist/measured-count drift in files this PR does not touch — `git diff upstream/main -- rust/processing/tests/module_size_allowlist.txt` is empty), SDK/sandbox/cli/mcp/viewer-core build, changeset added (minor bump on `@ifc-lite/sdk`, `@ifc-lite/sandbox`, `@ifc-lite/cli`, `@ifc-lite/mcp`, `@ifc-lite/viewer-core`).

Diff: 1148 insertions / 4 deletions across 20 files.

## What remains — stated, not filed as a sub-issue

Untouched by this PR, in the order #4510 proposed:

1. **Properties card** — a `StructuralCard` beside `ScheduleCard` in `PropertiesPanel.tsx`, listing a member's analysis model, connections and applied loads. Depends on this PR for the store slice.
2. **Geometry** — structural curve/surface members through a line channel (`setLineOverlay`-style), with a class toggle. `structural_analysis_curve.ifc`'s members carry `IfcTopologyRepresentation` over `IfcEdge`/`IfcVertexPoint`, not solids, and `processor_registry.rs` has no dispatch entry for that representation type — this is a new extraction path, not a toggle, and is the largest remaining piece.
3. **Write / round-trip** — a `structural-serializer.ts` matching `schedule-serializer.ts`, with a fixture-diff test. Nothing edits structural data today, so this has no caller yet; whether the goal is a round-trip test or the loss report in the "ends when" bar is still an open question raised in the issue comments.
4. **The second real-tool fixture** the issue's "ends when" bar asks for cannot be supplied by code: both other structural files in the manifest (`buildingsmart/Building-Structural.ifc`, `various/01_Snowdon_Towers_Sample_Structural(1).ifc`) were checked and carry zero structural-analysis entities (`IFCSTRUCTURALANALYSISMODEL`, `IFCSTRUCTURALCURVEMEMBER`, etc.) — they are structural products, not analysis models. `structural_analysis_curve.ifc` remains the corpus's only analysis fixture.
5. IFC2X3: not exercised against a real 2X3 structural export in this or the prior PR; #4510's docblock names the schema divergences it checked against the .exp grammar.

`stats-aggregation.ts` is untouched — the issue thread already established there is no exclusion there to remove.

## Queue gate

No closing keyword used deliberately — this delivers one of six layers and #4206 must stay open.

Refs #4206 (partial: layer 3 of the six the issue scopes; deliberately does not close it).

## Summary by CodeRabbit

* **New Features**
  * Added read-only structural analysis access through the `bim.structural` namespace.
  * Query analysis models, members, connections, activities, load groups, and result groups.
  * Structural data supports optional model selection and reports when load data was truncated.
  * Added typed structural analysis interfaces to the SDK.
* **Tests**
  * Added coverage for structural queries, data translation, model selection, caching, and truncation handling.

### CODE REVIEW

**Review by louistrue — CHANGES_REQUESTED** [review]
Independent review found two merge-blocking API-contract issues beyond the module-size regressions already noted:

1. packages/sdk/src/types.ts makes BimBackend.structural required. @ifc-lite/sdk is 5.0.0 and BimBackend is public; every external backend implementation that was valid before this additive feature now fails typechecking. The minor changeset therefore understates a breaking change. Preserve compatibility by making the new capability optional and have StructuralNamespace fail explicitly when an older backend does not provide it, or declare the required major bump.

2. The generated sandbox surface types every nested StructuralLoadConfiguration entry Value as unknown. Runtime translateConfiguration recursively emits a complete structural load, and the SDK type correctly models that recursion, but bim-globals.d.ts prevents script consumers from reading nested load components/configurations without unsafe assertions. Give the sandbox declaration a named recursive StructuralLoad shape (or equivalent fully typed recursion) and add a compile-level/template assertion that a nested Value exposes Components and Configuration.

The Ollama notes about numeric load components versus boolean-capable boundary conditions, and optional entry values, are not blockers: both exactly mirror the parser contracts. The former types different EXPRESS domains; the latter represents a dropped/unreadable Values slot and is paired with dropped.

**Review by louistrue — DISMISSED** [review]
The compatibility, generated recursive typing, module-size, and changeset blockers are resolved on `1673e79feed4ab7bc1be572434f99de4b6caf970`. The latest test-only commit is sound and covers all seven unsupported accessors directly in `namespaces.test.ts`; Ollama's remaining suggestion that `context.test.ts` repeat those same seven assertions is therefore redundant, not a defect. Structural load components are numeric by IFC schema, while boolean-capable boundary-condition components are a distinct object type.

Independent local typecheck/lint/package suites and exact-head Ollama are clean. Approving the code; it stays out of the merge queue until exact-head CI and the Claude/review-posted lane are green.

File: apps/viewer/src/sdk/adapters/structural-adapter.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  }
+  if (state.ifcDataStore) return state.ifcDataStore as IfcDataStore;
+  const activeId = state.activeModelId as string | null | undefined;
+  if (activeId) {
+    const active = getModelForRef(state, activeId);
+    if (active?.ifcDataStore) return active.ifcDataStore as IfcDataStore;
+  }
+  const firstFederated = state.models?.values().next().value;
+  return (firstFederated?.ifcDataStore as IfcDataStore | undefined) ?? null;
+}
```
[the remark this answers is not in the document]
Comment (BIMvoice) [reply apps/viewer/src/sdk/adapters/structural-adapter.ts:49]: Verified against the actual code: this is not a structural-only gap. structural-adapter.ts's resolveStore (lines 35-49) is byte-for-byte the same resolution order as schedule-adapter.ts's resolveStore (lines 29-46) -- active model, then first federated model, and a supplied-but-unresolvable modelId returns null -> EMPTY_EXTRACTION in both. Every other backend-local namespace (query, selection, viewer) follows the same pattern per the comment at schedule-adapter.ts:36-38. Making structural throw on an unknown modelId while every sibling namespace stays silent would introduce a new inconsistency rather than remove one. Not changing this in isolation here; a stricter unknown-model contract, if wanted, belongs in a follow-up that changes all these adapters together, not just structural's.

**Review by louistrue — APPROVED** [review]
Re-reviewed exact head 9b9627604. The only delta since the previously approved head is a compile-only template assertion for the recursive structural-load type; it is deliberately outside the user-facing template registry, uses the public generated BIM globals, and exact-head Ollama coverage completed. Prior implementation and validation findings remain resolved.
